src/ralph/backends/http/async_lrs.py

```diff
@@ -157,15 +160,20 @@ async def read(  # pylint: disable=too-many-arguments
 
         try:
             if greedy:
-                async for statement in self._greedy_fetch_statements(
+                statements_async_generator = self._greedy_fetch_statements(
                     target, raw_output, query_params
-                ):
-                    yield statement
+                )
             else:
-                async for statement in self._fetch_statements(
+                statements_async_generator = self._fetch_statements(
                     target=target, raw_output=raw_output, query_params=query_params
-                ):
-                    yield statement
+                )
+
+            counter = 0
+            async for statement in statements_async_generator:
+                if max_statements and (counter >= max_statements):
+                    break
+                yield statement
+                counter += 1
 
         except HTTPError as error:
             msg = "Failed to fetch statements."
```

Comment:
We should try to reduce as much as possible the size of a try/except block to make explicit which part of the code is under surveillance.

Here is an alternative (more concise) implementation:

```python
        statements_generator = self._fetch_statements
        if greedy:
            statements_generator = self._greedy_fetch_statements

        try:
            async for counter, statement in enumerate(statements_generator(target, raw_output, query_params)):
                if max_statements and (counter >= max_statements):
                    break
                yield statement
        except HTTPError as error:
            msg = "Failed to fetch statements."
            logger.error("%s. %s", msg, error)
            raise BackendException(msg, *error.args) from error
```

Reply:
oups. Fixed this !

Reply:
I would have preferred that as well but `enumerate` is apparently not compatible with async generators. (https://groups.google.com/g/comp.lang.python/c/stdbLogcUZo?pli=1)

Reply:
<img src="https://media3.giphy.com/media/W0c3xcZ3F1d0EYYb0f/giphy.gif"/>